### PR TITLE
Support mistral and sliding window attention

### PR DESCRIPTION
**Pull request by grimoire**
- [Mistral 7b v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1)
- sliding window attention kernel
- window block manager
- rename params

> [!IMPORTANT]
> This PR contains refactoring of the engine core mechanism.

### CODE REVIEW

File: lmdeploy/pytorch/models/mistral.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import torch
+import torch.distributed as dist
+from torch import nn
+from torch.distributed._tensor import DeviceMesh
+
+from ..dist_utils import (colwise_parallelize_linear_fn,
+                          rowwise_parallelize_linear_fn)
+from ..kernels.fill_kv_cache import fill_kv_cache
+from ..kernels.pagedattention import paged_attention_fwd
+from .llama import apply_rotary_pos_emb
```
Comment (RunningLeon) [inline lmdeploy/pytorch/models/mistral.py:13]: why not import `apply_rotary_pos_emb` from  `..kernels`?

**Comment by zhyncs** [comment]
@grimoire  is very productive, and the support for new models on the PyTorch engine is very timely. Currently, [Mistral](https://github.com/InternLM/lmdeploy/pull/1075), [Qwen 1.5](https://github.com/InternLM/lmdeploy/pull/1160), and [DeepSeek MoE](https://github.com/InternLM/lmdeploy/pull/1163) all rely on [sliding window attention](https://github.com/InternLM/lmdeploy/pull/1075). @lvhan028 @RunningLeon Will we consider prioritizing this PR for review and merging it as soon as possible? Given that these are relatively large features, the community would likely appreciate being able to use them sooner rather than later.

**Comment by RunningLeon** [comment]
> @grimoire is very productive, and the support for new models on the PyTorch engine is very timely. Currently, [Mistral](https://github.com/InternLM/lmdeploy/pull/1075), [Qwen 1.5](https://github.com/InternLM/lmdeploy/pull/1160), and [DeepSeek MoE](https://github.com/InternLM/lmdeploy/pull/1163) all rely on [sliding window attention](https://github.com/InternLM/lmdeploy/pull/1075). @lvhan028 @RunningLeon Will we consider prioritizing this PR for review and merging it as soon as possible? Given that these are relatively large features, the community would likely appreciate being able to use them sooner rather than later.

@zhyncs Hi, feel free to review and test this PR. Any comment would be sincerely appreciated.

**Comment by lvhan028** [comment]
This PR makes a great change. QA needs more time to test it.
I am afraid it cannot catch the latest v0.2.4 version, which will be released this week. 
We'll release it in v0.2.5 next week.

File: lmdeploy/model.py [hunk 2]
```diff
@@ -56,6 +56,7 @@ def chat_template(self):
         return model
 
 
+@MODELS.register_module(name='mistral')
```
Comment (lvhan028) [inline lmdeploy/model.py:59]: What's the chat template?

File: docs/en/supported_models/supported_models.md [hunk 3]
```diff
@@ -28,3 +28,4 @@
 | ChatGLM2  |    6B     |    Yes    |   No    |  No  |
 |  Falcon   | 7B - 180B |    Yes    |   No    |  No  |
 |    YI     | 6B - 34B  |    Yes    |   No    |  No  |
+|  Mistral  |    7B     |    Yes    |   No    |  No  |
```
Comment (RunningLeon) [inline docs/en/supported_models/supported_models.md:31]: Could update the readme as well:
https://github.com/InternLM/lmdeploy/blob/main/README.md#supported-models

File: lmdeploy/model.py (hunk 2 above)
Comment (grimoire) [reply lmdeploy/model.py:59]: I don't know. Mistral does not provide a chat model.

Comment (RunningLeon) [reply lmdeploy/model.py:59]: Mistral has models like [mistralai/Mistral-7B-Instruct-v0.1](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1),  [mistralai/Mistral-7B-Instruct-v0.2](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2).
I write the [chat template](https://github.com/InternLM/lmdeploy/blob/25e905ffd41e6c26253aceff620c16e228acc9db/lmdeploy/model.py#L943) in PR #1133. Could you check and review the PR?

**Comment by lvhan028** [comment]
@RunningLeon Is there evaluation result of this PR?

File: lmdeploy/model.py (hunk 2 above)
Comment (lvhan028) [reply lmdeploy/model.py]: can the chat template in #1133 be merged into this one?

**Comment by RunningLeon** [comment]
> @RunningLeon Is there evaluation result of this PR?

here are the results. There might be something wrong.
13dde2c12e34d89a4b9aeeb7b25aec4

File: lmdeploy/pytorch/paging/eviction_helper/recompute_eviction_helper.py [hunk 4]
```diff
@@ -1,16 +1,19 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from typing import Dict
 
-from lmdeploy.pytorch.paging.block_manager import BlockManager
-
 from ...messages import SchedulerSequence
 from .base_eviction_helper import BaseEvictionHelper
 
 
+def _div_up(x, n):
```
Comment (lvhan028) [inline lmdeploy/pytorch/paging/eviction_helper/recompute_eviction_helper.py:8]: useless function?

**Review by lvhan028 — APPROVED** [review]

**Comment by lvhan028** [comment]
@zhulinJulia24 may perform regression test.

**Comment by RunningLeon** [comment]
@lvhan028 The perform test is OK as here

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**